### PR TITLE
Cosmetic filtering

### LINKED ISSUE #3: Implement cosmetic blocking [tracking]
Ongoing work in `adblock-rust` to implement cosmetic blocking will need new APIs exposed through `adblock-rust-ffi` to be useful for the Brave browser.

### PR DESCRIPTION
**Pull request by antonok-edm**
Closes #3, depends on https://github.com/brave/adblock-rust/pull/47

### CODE REVIEW

File: examples/cpp/main.cpp [hunk 1]
```diff
@@ -231,6 +231,56 @@ void TestThirdParty() {
...
+      "##.block\n"
+      "a.com#@#.block\n"
+      "##a[href=\"b.com\"]\n"
+      "b.*##div:style(background: #fff)\n"
+  );
+
+  std::string a_resources = engine.hostnameCosmeticResources("a.com");
+  std::string a_order1("{\"hide_selectors\":[\"a[href=\\\"b.com\\\"]\",\"#element\"],\"style_selectors\":{},\"exceptions\":[\".block\"],\"injected_script\":\"\"}");
+  std::string a_order2("{\"hide_selectors\":[\"#element\",\"a[href=\\\"b.com\\\"]\"],\"style_selectors\":{},\"exceptions\":[\".block\"],\"injected_script\":\"\"}");
+  assert(a_resources == a_order1 || a_resources == a_order2);
```
Comment (bbondy) [inline examples/cpp/main.cpp]: Just curious, order is not guaranteed to be deterministic?

File: examples/cpp/main.cpp [hunk 2]
```diff
@@ -231,6 +231,56 @@ void TestThirdParty() {
...
+  assert(stylesheet == ".ads,#element{display:none !important;}");
+
+  classes = std::vector<std::string>({"element", "a"});
+  ids = std::vector<std::string>({"block", "ads", "a"});
+  exceptions = std::vector<std::string>({"#block"});
+  stylesheet = engine.classIdStylesheet(classes, ids, exceptions);
+  assert(stylesheet == ".element,#block + .child,#ads > #element{display:none !important;}");
+}
+
+void TestHostnameCosmetics() {
```
Comment (bbondy) [inline examples/cpp/main.cpp]: @antonok-edm could you also add a test with a subdomain to show what the expected usage is w.r.t. doing eTLD+1 or not.

File: examples/cpp/main.cpp [hunk 3]
```diff
@@ -231,6 +231,56 @@ void TestThirdParty() {
...
+  std::vector<std::string> ids = std::vector<std::string>();
+  std::vector<std::string> exceptions = std::vector<std::string>();
+  std::string stylesheet = engine.classIdStylesheet(classes, ids, exceptions);
+  assert(stylesheet == "");
+
+  classes = std::vector<std::string>({"ads", "no-ads"});
+  ids = std::vector<std::string>({"element"});
+  exceptions = std::vector<std::string>();
+  stylesheet = engine.classIdStylesheet(classes, ids, exceptions);
+  assert(stylesheet == ".ads,#element{display:none !important;}");
```
Comment (bbondy) [inline examples/cpp/main.cpp]: Could do a couple more tests too for invalid input and for something like passing in IDs with `#` showing that it doesn't work in that case.

File: examples/cpp/main.cpp [hunk 4]
```diff
@@ -231,6 +231,56 @@ void TestThirdParty() {
...
+      "a.com###element\n"
+      "b.com##.ads\n"
+      "##.block\n"
+      "a.com#@#.block\n"
+      "##a[href=\"b.com\"]\n"
+      "b.*##div:style(background: #fff)\n"
+  );
+
+  std::string a_resources = engine.hostnameCosmeticResources("a.com");
+  std::string a_order1("{\"hide_selectors\":[\"a[href=\\\"b.com\\\"]\",\"#element\"],\"style_selectors\":{},\"exceptions\":[\".block\"],\"injected_script\":\"\"}");
```
Comment (bbondy) [inline examples/cpp/main.cpp]: Could you add some tests for `injected_script` too?

File: examples/cpp/main.cpp (hunk 1 above)
Comment (antonok-edm) [reply examples/cpp/main.cpp]: Nope, these are pulled straight from `HashMap`s in the standard Rust library. Order of iteration is non-deterministic. Could probably choose another data structure if that's desired.

Comment (bbondy) [reply examples/cpp/main.cpp]: Non-deterministic is ok, just checking. Thanks.

**Pull request merged**